src/github.com/couchbase/sync_gateway/db/kv_change_index.go

```diff
@@ -996,3 +973,243 @@ func (db *DatabaseContext) singleChannelStats(kvIndex *kvChangeIndex, channelNam
 func IsNotFoundError(err error) bool {
 	return strings.Contains(strings.ToLower(err.Error()), "not found")
 }
+
+type unmarshalWorker struct {
+	output  chan<- *LogEntry
+	processing chan *unmarshalEntry
+}
+
+
+// UnmarshalWorker is used to unmarshal incoming entries from the DCP feed in parallel, while maintaining ordering of
+// sequences per vbucket.  The main DocChanged loop creates one UnmarshalWorker per vbucket.
+func NewUnmarshalWorker(output chan<- *LogEntry) *unmarshalWorker {
+
+	// goroutine to work the processing channel and return completed entries
+	// to the output channel
+	maxProcessing := 500
+
+	worker := &unmarshalWorker{
+		output: output,
+		processing: make(chan *unmarshalEntry, maxProcessing),
+	}
+
+	// Start goroutine to work the worker's processing channel.  When an entry arrives on the
+	// processing channel, blocks waiting for success/fail for that document.
+	go func() {
+		for {
+			select {
+				case unmarshalEntry:= <-worker.processing:
+					// Wait for entry processing to be done
+					entryTime := time.Now()
+					select {
```

Comment:
I don't think this nested `select` adds any value, since there's only a single branch and no default branch.